### PR TITLE
Implemented BackpropGetArcPost and fixed some issues of ArgMax

### PR DESCRIPTION
**Pull request by qindazhu**
- Implemented BackpropGetArcPost (will add tests later, merge now to avoid future conflicts as there are many style changes)

- Fixed some issues of `ArgMaxPerSublist`
- Refactor testing code of `ArgMaxPerSublist`

### CODE REVIEW

**Comment by qindazhu** [comment]
Please ignore `ragged.h` and `ragged_ops.cu` for review, they are just style issue fixes (mainly for adding `explicit` to constructor).

File: k2/csrc/ragged_ops.h [hunk 1]
```diff
@@ -81,25 +82,21 @@ void LogSumPerSublist(Ragged<T> &src, T initial_value, Array1<T> *dst_values) {
...
-                                >= 2. src.values is allowed to be empty.
-     @param [out] argmax        Array to which the argmax indexes will be written.
-                                max_values->Dim() == src.TotSize(src.NumAxes() - 1).
+  if the sub-list was empty or all values in the sub-list are less than
+  `initial_value`.
+
+     @param [in] src        Input ragged array; must have src.NumAxes() >= 2.
+                            src.values is allowed to be empty.
+     @param [out] argmax    Array to which the argmax indexes will be written.
+                            max_values->Dim() == src.TotSize(src.NumAxes() - 2),
```
Comment (csukuangfj) [inline k2/csrc/ragged_ops.h]: There is no `max_values` here.

File: k2/csrc/ragged_ops_inl.h [hunk 2]
```diff
@@ -485,48 +473,47 @@ struct PairOutputIteratorDeref {  // this is what you get when you dereference
...
-  __device__ Pair<T> operator() (const Pair<T> &a,
-                                 const Pair<T> &b) const {
+  __device__ __forceinline__ Pair<T> operator()(const Pair<T> &a,
+                                                const Pair<T> &b) const {
     // NOTE: could specialize this via a union, if T == int32_t, might be
     // marginally faster.
     if (a.t > b.t || (a.t == b.t && a.idx > b.idx))
       return a;
-    else return b;
+    else
```
Comment (csukuangfj) [inline k2/csrc/ragged_ops_inl.h]: `else` can be removed.

File: k2/csrc/ragged_ops.h (hunk 1 above)
Comment (qindazhu) [reply k2/csrc/ragged_ops.h]: Thanks, Fixed

**Pull request closed without merging**